va/http.go

```diff
@@ -0,0 +1,455 @@
+			TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
+			// We don't expect to make multiple requests to a client, so close
+			// connection immediately.
+			DisableKeepAlives: true,
+			// We don't want idle connections, but 0 means "unlimited," so we pick 1.
+			MaxIdleConns:        1,
+			IdleConnTimeout:     time.Second,
+			TLSHandshakeTimeout: 10 * time.Second,
+		},
+		CheckRedirect: checkRedirect,
+	}
+}
+
+// httpValidationURL constructs a URL for the given IP address, path and port
+// combination. The port is omitted from the URL if it is the default HTTP
+// port. It is only used for constructing initial HTTP-01 validation requests
+// and so does not need to support constructing a URL with an HTTPS scheme.
+func httpValidationURL(validationIP net.IP, path string, port int) *url.URL {
+	urlHost := validationIP.String()
+
+	// If the port is something other than the conventional HTTP port, put it in
+	// the URL.
+	if port != 80 {
+		urlHost = net.JoinHostPort(validationIP.String(), strconv.Itoa(port))
+	}
+
+	// if the validation IP is an IPv6 address, and we aren't using
+	// `net.JoinHostPort` then we have to manually surround the IPv6 address
+	// with square brackets to make a valid IPv6 URL (e.g "http://[::1]/foo" not
+	// "http://::1/foo")
+	if port == 80 && validationIP.To4() == nil {
+		urlHost = fmt.Sprintf("[%s]", urlHost)
+	}
+
+	return &url.URL{
+		Scheme: "http",
+		Host:   urlHost,
+		Path:   path,
+	}
+}
+
+// httpValidationTarget bundles all of the information needed to make an HTTP-01
+// validation request against a target.
+type httpValidationTarget struct {
+	// the hostname being validated
+	host string
+	// the port for the validation request
+	port int
+	// the path for the validation request
+	path string
+	// all of the IP addresses available for the host
+	available []net.IP
+	// the IP addresses that were tried for validation previously that were cycled
+	// out of cur by calls to nextIP()
+	tried []net.IP
+	// the IP addresses that will be drawn from by calls to nextIP() to set curIP
+	next []net.IP
+	// the current IP address being used for validation (if any)
+	cur *net.IP
+}
+
+// nextIP changes the cur IP by removing the first entry from the next slice and
+// setting it to cur. If cur was previously set the value will be added to the
+// tried slice to keep track of IPs that were previously used. If nextIP() is
+// called but vt.next is empty an error is returned.
+func (vt *httpValidationTarget) nextIP() error {
+	if len(vt.next) == 0 {
+		return fmt.Errorf(
+			"host %q has no IP addresses remaining to use",
+			vt.host)
+	}
+	oldIP := vt.cur
+	if oldIP != nil {
+		vt.tried = append(vt.tried, *oldIP)
+	}
+	vt.cur = &vt.next[0]
+	vt.next = vt.next[1:]
+	return nil
+}
+
+// ip returns the current *net.IP for the validation target. It may return nil
+// if all possible IPs have been expended by calls to nextIP.
+func (vt *httpValidationTarget) ip() *net.IP {
+	return vt.cur
```

Comment:
Any particular reason for returning the pointer here instead of returning the derefed net.IP? It seems like the main the caller ends up needing to deref it themselves anyway.

Reply:
Nope! Optimistically there was a reason it made sense before the last refactor but it doesn't make sense now. f6509f2 should do the trick.